federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,244 @@ limitations under the License.
+				}
+			}
+		}
+	}
 	return nil
 }
+
+/* ensureDnsRecords ensures (idempotently, and with minimum mutations) that all of the DNS records for a service in a given cluster are correct, given the current state of that service in that cluster.  This should be called every time the state of a service might have changed (either w.r.t. it's loadblancer address, or if the number of healthy backend endpoints for that service transitioned from zero to non-zero (or vice verse).  Only shards of the service which have both a loadbalancer ingress IP address or hostname AND at least one healthy backend endpoint are included in DNS records for that service (at all of zone, region and global levels). All other addresses are removed.  Also, if no shards exist in the zone or region of the cluster, a CNAME reference to the next higher level is ensured to exist.
+ */
+func (s *ServiceController) ensureDnsRecords(clusterName string, cachedService *cachedService) error {
+	// Quinton: Pseudocode....
+	// See https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648
+	// For each service we need the following DNS names:
+	// mysvc.myns.myfed.svc.z1.r1.mydomain.com  (for zone z1 in region r1)
+	//         - an A record to IP address of specific shard in that zone (if that shard exists and has healthy endpoints)
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.r1.mydomain.com  (if a healthy shard does not exist in zone z1)
+	// mysvc.myns.myfed.svc.r1.federation
+	//         - a set of A records to IP addresses of all healthy shards in region r1, if one or more of these exist
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.mydomain.com (if no healthy shards exist in region r1)
+	// mysvc.myns.myfed.svc.federation
+	//         - a set of A records to IP addresses of all healthy shards in all regions, if one or more of these exist.
+	//         - no record (NXRECORD response) if no healthy shards exist in any regions)
+	//
+	// For each cached service, cachedService.lastState tracks the current known state of the service, while cachedService.appliedState contains
+	// the state of the service when we last successfully sync'd it's DNS records.
+	// So this time around we only need to patch that (add new records, remove deleted records, and update changed records.
+	//
+	if s == nil {
+		return fmt.Errorf("nil ServiceController passed to ServiceController.ensureDnsRecords(clusterName: %s, cachedService: %v)", clusterName, cachedService)
+	}
+	if cachedService == nil {
+		return fmt.Errorf("nil cachedService passed to ServiceController.ensureDnsRecords(clusterName: %s, cachedService: %v)", clusterName, cachedService)
+	}
+	serviceName := cachedService.lastState.Name
+	namespaceName := cachedService.lastState.Namespace
+	if zoneNames, regionName, err := s.getClusterZoneNames(clusterName); err != nil {
+		return err
+	} else {
+		if dnsZoneName, err := s.getFederationDNSZoneName(); err != nil {
```

Comment:
So here is the convention you can see in most idiomatic Go code.

##### Case 1
```Go
v, err := fValErr()
if err != nil {
    return fmt.Errorf("Message explaining the reason for failure at this point: %v", err)
}
use(v)   // effectively else block but dedented to make it easier to keep track of the flow.
```

##### Case 2
```Go
if err := fErrOnly(); err != nil {
    return fmt.Errorf("Message explaining the reason for failure at this point: %v", err)
}
// Function does not return any value, so there is no "use(v)" 
```

##### Case 3
```Go
if _, err := fValErr(); err != nil {
    return fmt.Errorf("Message explaining the reason for failure at this point: %v", err)
}
// Function does return value v, but we don't care about it because there is no "use(v)" 
```

Makes sense?